### ISSUE TITLE
Rework the defaults system to be more flexible with custom plist

### ISSUE THREAD

**Issue by entusdictus** [aonez/Keka#15]
When trying to unzip a zip file that was previously created with the macOS Sierra integrated compressor, Keka fails with "Error Code 2 using p7zip" - Fatal Error. 

Keka removes the .zip file extension and leaves a plain file in place. 

Using the integrated compressor to unzip the same file works as expected.

**Comment by benungs** [comment]
I received the same error today, on OS X El Capitan. It was a huge file, 9.42GB, and was uploaded on a mac with Dropbox (presumably zipped OS X native, but that's a guess), downloaded via Firefox. Same as entusductus, I was able to unzip it with the OS X's native archive utility.  I tried it with Keka version 1.0.7 (latest stable release) and the beta version (1.1.0b1r1827), both with the same 'Error Code 2 using "p7zip" Fatal error' message.

![screen shot 2017-01-27 at 2 01 00 pm](https://cloud.githubusercontent.com/assets/25394210/22387487/11da9ed2-e499-11e6-8ec6-da98ff0d4c86.jpg)

**Comment by aonez** [comment]
Can any of those files be shared so I can test?

**Comment by entusdictus** [comment]
sure, as this is reproducible with any newly created zip file, sharing a file is no problem. 
[testfile.txt.zip](https://github.com/aonez/Keka/files/738261/testfile.txt.zip)

**Comment by parkerbr** [comment]
any solutions yet??

**Comment by aonez** [comment]
@entusdictus I'm having no problem with the attached file. Can you delete the Keka preferences file and try again? The file is located here:

`/Users/YOURUSERNAME/Library/Preferences/com.aone.keka.plist`

@parkerbr, note that the error 2 is generic and may be caused for different reasons, you have any trouble with the attached file?

@benungs, can you try with this older version: [Keka 1.0.4](http://download.kekaosx.com/snow/Keka-1.0.4-snow.dmg)

**Comment by entusdictus** [comment]
@aonez deleting the preferences worked. It seems that there are missing keys in the plist. 
FYI, we put a default plist in the user's home that turns off auto updates. Maybe Keka has a problem when some keys already exist in the plist and this causes other key not to be written.
I've attached three files for you.

com.aone.keka created with initial file.plist -> pre-created plist as base, Keka filled the remaining keys
com.aone.keka new created.plist -> the plist that was created without the pre-created file
com.aone.keka.plist -> the pre-created file that is used in a default user template

There are differences between the two plists - "created with initial file" and "new created". There are some missing keys within the "created with initial file".

[plist files.zip](https://github.com/aonez/Keka/files/738779/plist.files.zip)

**Comment by aonez** [comment]
@entusdictus its not the first time the plist gets corrupted. I don't understand clearly your procedure. Correct me if I'm wrong:

1. Add custom `com.aone.keka.plist`, containing some customs keys.
2. Open Keka, it merges and creates the file you've called `com.aone.keka created with initial file.plist`

Is that right?

**Comment by benungs** [comment]
@aonez Deleting prefs didn't work, with any version (same error each time). The 1.0.4 version you posted failed sooner than 1.0.7, but actually produced a folder, though only a few of the items actually unzip correctly; the rest are empty:

![screen shot 2017-01-30 at 8 10 44 am](https://cloud.githubusercontent.com/assets/25394210/22428274/a157e4b2-e6c3-11e6-91d6-2bdc23962b6c.jpg)

The custom plist files didn't help either. 
Again, these unzip properly with the native OS X archive utility. I think it's a size issue. I'm going to run some tests with different file sizes.

**Comment by benungs** [comment]
It looks like Keka struggles with zipped files over 4GB. Here you can find two files zipped with OS X's Archive Utility; one slightly over 4GB, the other slightly under 4GB. The smaller one unzips; the larger one fails. 

http://mstly.work/keka

**Comment by entusdictus** [comment]
@aonez 100% correct :)

**Comment by aonez** [comment]
@benungs your issue will be issued here: #18

**Comment by aonez** [comment]
Will have to rework the defaults read/write system.

@entusdictus If you just want to disable the auto-updates and the pop up to enable them, you can do it by adding the plist with the options you've already added, but just those properties:

```
SUEnableAutomaticChecks = false
SUHasLaunchedBefore = true
SUSendProfileInfo = false
```

Any other property (like `ExcludeMacForks`) will be override if you don't set the version in the plist. If you want to set other properties, be careful to add all the defaults for the specific version also, otherwise Keka will be broken. I recommend you to set up Keka with your wanted defaults and then edit/share the plist Keka creates.

Anyway I'll remake (sometime) the defaults system so it is more flexible.

**Comment by entusdictus** [comment]
Thanks @aonez, setting only those three keys works. Thanks for your assistance! 👍

**Comment by ejb4u** [comment]
Hi Aonez,
I still have "error code 2 using p7zip" with Keka 1.1.3 after some tests with pre-configured plists before install. Even if I remove all files in ~/Library/Containers and reinstall Keka 1.1.3 (downloaded from the website) it never work now.
I'm digging into [the official Apple documentation](https://developer.apple.com/library/archive/documentation/Security/Conceptual/AppSandboxDesignGuide/AboutAppSandbox/AboutAppSandbox.html#//apple_ref/doc/uid/TP40011183-CH1-SW1) to better understand.
I'm an Xcode user and work with Swift or Objective-C.
How could I help you ?

**Comment by aonez** [comment]
Hi @ejb4u, be sure to delete all:

- `Keka.app`
- `~/Library/Containers/com.aone.keka` folder
- `~/Library/Preferences/com.aone.keka`

**Comment by ejb4u** [comment]
Thanks for the reply.
I just didi it again, and oops still the same error.
The container is probably storing data elsewhere. I'm searching and will keep you informed.
Nevertheless the zip file that I want to unzip is **deleted** when there is the error message, so I don't get it extracted, furthermore I lose the original zip file. In Keka preferences nothing is ticked so the zip file should not be removed. This should also be fixed...
And when asking for the preferences, the first time nothing happens, I must ask for the preferences twice to see the panel which is empty. Probably a container problem.
MacOs 10.13.6 & Keka 1.1.3 (2809)

**Comment by aonez** [comment]
@ejb4u there's no more data of Keka anywhere else... Be sure to delete de **folder** in the sandbox container. Maybe a restart will help cleaning caches?

> the zip file that I want to unzip is deleted when there is the error message

This has been reported multiple times, but I've never reproduced it 😞 If you can, share those custom plists so I can tests.

> And when asking for the preferences, the first time nothing happens...

This reminds me of #226, that's why I asked to be sure to delete the sandbox folder. Take a look if you also see the old toolbar instead of the unified.

**Comment by ejb4u** [comment]
Yes this is exactly as #226 but I always remove the whole directory. I don't know how to solve the issue.

I've found other places with Keka files in /var, but erasing them doesn't solve the issue.

```
sudo find -L /var -name "*aone*" -exec echo file:{} \; -exec ls -l {} \; 2>/dev/null
file:/var/db/BootCaches/02FB2422-542A-4644-86F1-B2CC56D4BCA3/app.com.aone.keka.playlist
-rw-------  1 root  wheel  71516 27 aoû 09:36 /var/db/BootCaches/02FB2422-542A-4644-86F1-B2CC56D4BCA3/app.com.aone.keka.playlist
file:/var/db/BootCaches/PreheatedUser/app.com.aone.keka.playlist
-rw-------  1 root  wheel  71516 27 aoû 09:36 /var/db/BootCaches/PreheatedUser/app.com.aone.keka.playlist
file:/var/folders/0b/xk88km593491qjpb65dn5wmc0000gn/T/com.aone.keka
file:/var/folders/0b/xk88km593491qjpb65dn5wmc0000gn/C/com.aone.keka
total 0
drwxr-xr-x@ 3 myuser  staff   96 23 aoû 16:56 com.aone.keka
drwxr-xr-x  5 myuser  staff  160 23 aoû 16:58 com.apple.metal
file:/var/folders/0b/xk88km593491qjpb65dn5wmc0000gn/C/com.aone.keka/com.aone.keka
total 0
drwxr-xr-x@ 5 myuser  staff  160 23 aoû 16:56 com.apple.metal
```

**Comment by aonez** [comment]
This should be fixed in the latests 1.2.0 development builds.
